File: 2024/day2.js
const { on } = require('cluster');

const fs = require('fs').promises;

let readInput = async () => {
    let res = await fs.readFile('./../input.txt');
    res = res.toString().split('\n')
        .map(line => line.split(/\s+/).map(Number));
    // console.log(res);
    return res;

}

let findSafeReports = async (inputs) => {
    // input array of 2 number pairs

    const isIncreasing = (org) => {
        let arr = [...org]
        for (let i = 1; i < arr.length; i++) {
            if (arr[i] <= arr[i - 1]) { // 
                return false;
            }
            const dist = arr[i] - arr[i - 1];
            if (dist > 3) {
                return false;
            }
            return true;
        }
    }

    const isDecreasing = (arr) => {
        for (let i = 1; i < arr.length; i++) {
            if (arr[i] >= arr[i - 1]) {
                return false;
            }
            const dist = arr[i - 1] - arr[i];
            if (dist > 3) {
                return false;
            }
        }
        return true;
    }

    let count = 0;
    for (const report of inputs) {
        if (isDecreasing(report) || isIncreasing(report)) {
            count++;
        }
    }

    return count
}

// 1 2 3 4 6 5 9 10 

let main = async () => {
    let inputs = await readInput();
    let sum = await findSafeReports(inputs);
    console.log(sum);
}

main();

File: 2024/day2.py
import asyncio

async def read_input():
    with open('./input.txt', 'r') as f:
        res = f.read().strip().split('\n')
        res = [list(map(int, line.split())) for line in res]
    return res

def is_increasing(arr):
    for i in range(1, len(arr)):
        if arr[i] <= arr[i-1]:
            return False
        dist = arr[i] - arr[i-1]
        if dist > 3:
            return False
    return True

def is_decreasing(arr):
    for i in range(1, len(arr)):
        if arr[i] >= arr[i-1]:
            return False
        dist = arr[i-1] - arr[i]
        if dist > 3:
            return False
    return True

def combinations(arr):
    combs = []
    for i in range(len(arr)):
        comb = arr[:i] + arr[i+1:]
        combs.append(comb)
    return combs

async def find_safe_reports(inputs):
    count = 0
    for report in inputs:
        if is_decreasing(report) or is_increasing(report):
            count += 1
        else:
            combs = combinations(report)
            for comb in combs:
                if is_decreasing(comb) or is_increasing(comb):
                    count += 1
                    break
    return count

async def main():
    inputs = await read_input()
    sum_val = await find_safe_reports(inputs)
    print(sum_val)
    print("done")

if __name__ == "__main__":
    asyncio.run(main())
